ServerGUI.py

```diff
@@ -66,20 +87,106 @@ def create_widgets(self):
     def main_loop(self):
         """
         """
-        print('running main loop')
         #  TODO 1: update location based on indoor GPS
         self.update_loc()
         self.drawC1C0()
+        self.server.send_update((self.curr_tile.row, self.curr_tile.col))
         #  TODO 2: Update environment based on sensor data
-        print('server data, ', self.server.receive_data())
-        #  TODO 3: check if the previous path is obstructed
-        # If valid continue execution
-        # else re-plan path
-        # TODO 4: Send movement command
-        # TODO 5: return if we are at the end destiation
-        # loop
+        self.sensor_state = self.server.receive_data()
+        if type(self.sensor_state) is SensorState.SensorState:
+            self.visibilityDraw(self.sensor_state.lidar)
+            if self.grid.update_grid_tup_data(self.curr_tile.x, self.curr_tile.y, self.sensor_state.lidar, Tile.lidar, robot_radius, bloat_factor, self.path_set):
+                self.path = search.a_star_search(self.grid, (0, 0), self.endPoint, search.euclidean)
+                self.path = search.segment_path(self.grid, self.path)
+                self.path_set = set()
+                for tile in self.path:
+                    self.path_set.add(tile)
+            self.drawPath()
+        else:
+            print(self.sensor_state)
+            print('Ensure that a client thread has been started and is sending sensor data!')
+        self.calcVector()
+        if self.curr_tile == self.path[self.pathIndex]:
+            self.pathIndex += 1
+        # return if we are at the end destination
+        if self.curr_tile == self.path[-1]:
+            return
+        # recursively loop
         self.master.after(1, self.main_loop)
 
+    def calc_dist(self):
+        """
+        returns the perpendicular distance from c1c0's current value to the line that c1c0 should be travelling on
+        """
+        #(y2-y1)x-(x2-x1)y=(y2-y1)x1-(x2-x1)y1
+        #C = x2y1-x1y2
+        b = self.path[self.pathIndex + 1].x - self.path[self.pathIndex].x
+        a = self.path[self.pathIndex + 1].y - self.path[self.pathIndex].y
+        #c = self.path[self.pathIndex + 1].x * self.path[self.pathIndex].y \
+        #    - self.path[self.pathIndex].x * self.path[self.pathIndex + 1].y
+        den = 0
+        if a != 0 and b != 0:
+            den = (b/a) + (a/b)
+        d = 0
+        if den != 0:
+            c = self.path[self.pathIndex].y - (a/b)*self.path[self.pathIndex].x
+            x = (self.curr_tile.x + (b/a)*self.curr_tile.x - self.path[self.pathIndex].y + (a/b) *
+                 self.path[self.pathIndex].x)/den
+            y = (a/b)*x + c
+            d = ((x - self.curr_tile.x)**2 + (y - self.curr_tile.y)**2)**(1/2)
+        return d
+
+
+    def PID(self):
```

Comment:
@ShivalikaGupta  Hey could we make PID a class...? Then take some code out of here and the other files...

Reply:
I think it would clean things up and avoid duplication of controls code